src/neoxp/Node/NodeUtility.cs

```diff
@@ -138,5 +142,127 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_Contract).Add(state.Hash);
+            var localContract = snapshot.GetAndChange(key)?.GetInteroperable<ContractState>();
+            
+            if (localContract is null)
+            {
+                // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+                // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+                // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+                // over keys that shouldn't exist for one of the contracts.
+                if (state.Id < LastUsedContractId(snapshot))
+                {
+                    state.Id = GetNextAvailableId(snapshot);
+                }
+                else
+                {
+                    // Update available id such that a regular contract deploy will use the right next id;
+                    SetLastUsedContractId(snapshot, state.Id);
+                }
+                snapshot.Add(key, new StorageItem(state));
+            }
+            else 
+            {
+                if (force == ContractCommand.OverwriteForce.None)
+                {
+                    throw new Exception("Contract already exists locally. Use --force:<option> to overwrite");
+                }
+                if (force == ContractCommand.OverwriteForce.All ||
+                    force == ContractCommand.OverwriteForce.ContractOnly)
+                {
+                    // Note: a ManagementContract.Update() will not change the contract hash. Not even if the NEF changed.
+                    localContract.Nef = state.Nef;
+                    localContract.Manifest = state.Manifest;
+                    localContract.UpdateCounter = state.UpdateCounter;
+                }
+            }
+
+            if (force == ContractCommand.OverwriteForce.All || force == ContractCommand.OverwriteForce.StorageOnly)
+            {
+                var contractId = localContract is null ? state.Id : localContract.Id;
+
+                // the storage schema might have changed therefore we first clear all storage
+                byte[] prefix_key = StorageKey.CreateSearchPrefix(contractId, new byte[]{});
+                foreach (var (k, v) in snapshot.Find(prefix_key))
+                {
+                    snapshot.Delete(k);
+                }
+
+                for (int i = 0; i < storagePairs.Count; i++)
+                {
+                    snapshot.Add(
+                        new StorageKey { Id = contractId, Key = Convert.FromBase64String(storagePairs[i].key) },
+                        new StorageItem(Convert.FromBase64String(storagePairs[i].value)));
+                }
```

Comment:
It's been my understanding that for loops are slightly more performant than foreach for indexed types. foreach creates an enumerator object while indexed access does not. But it's been a long time since I learned that. I will do some research on the current best practices. Let's leave it for loop for now to be consistent with the rest of the code base.